confdb/confdb.go

```diff
@@ -1002,66 +1010,111 @@ func schemaTypesStr(types []SchemaType) string {
-	// we have the full storage path and final value
-	if placeIndex == len(unmatchedSuffix) {
-		return map[string]any{storagePath: val}, nil
-	}
+		case mapKeyType:
+			// use the non-placeholder parts of the suffix to find the value to write
+			mapVal, ok := val.(map[string]any)
+			if !ok {
+				return nil, fmt.Errorf(`expected map for unmatched request parts but got %T`, val)
+			}
 
-	mapVal, ok := val.(map[string]any)
-	if !ok {
-		return nil, fmt.Errorf(`expected map for unmatched request parts but got %T`, val)
-	}
+			val, ok = mapVal[unmatchedPart.name()]
+			if !ok {
+				return nil, fmt.Errorf(`cannot use unmatched part %q as key in %v`, unmatchedPart, mapVal)
+			}
 
-	storagePathsToValues := make(map[string]any)
-	// suffix has an unmatched placeholder, try all possible values to fill it and
-	// find the corresponding nested value.
-	for cand, candVal := range mapVal {
-		newStoragePath := replaceIn(storagePath, unmatchedSuffix[placeIndex].access(), cand)
-		pathsToValues, err := getValuesThroughPathsImpl(newStoragePath, unmatchedSuffix[placeIndex+1:], candVal)
-		if err != nil {
-			return nil, err
-		}
+		case indexPlaceholderType:
+			list, ok := val.([]any)
+			if !ok {
+				return nil, fmt.Errorf(`expected list for unmatched request parts but got %T`, val)
+			}
 
-		for path, val := range pathsToValues {
-			storagePathsToValues[path] = val
+			// TODO: can this be optimised? Maybe by changing the databag logic to be more
+			// match-aware instead of using these values to expand the matches?
+			storagePathsToValues := make(map[string]any)
+			for i, el := range list {
+				newStoragePath, err := replaceIn(storagePath, unmatchedPart.access(), "["+strconv.Itoa(i)+"]")
+				if err != nil {
+					return nil, err
+				}
+
+				pathsToValues, err := getValuesThroughPathsImpl(newStoragePath, unmatchedSuffix[unmatchedIndex+1:], el)
+				if err != nil {
+					return nil, err
+				}
+
+				for path, val := range pathsToValues {
+					storagePathsToValues[path] = val
+				}
+
+				storagePathsToValues[newStoragePath] = el
+			}
+			return storagePathsToValues, nil
+
+		case listIndexType:
+			list, ok := val.([]any)
+			if !ok {
+				return nil, fmt.Errorf(`expected list for unmatched request parts but got %T`, val)
+			}
+
+			index, _ := strconv.Atoi(unmatchedPart.name())
+			if index >= len(list) {
+				return nil, fmt.Errorf(`cannot use unmatched part %q as key in %v`, unmatchedPart.access(), list)
+			}
+
+			val = list[index]
 		}
 	}
 
-	return storagePathsToValues, nil
+	// we reached the end of the suffix (there are no unmatched placeholders) so
+	// we have the full storage path and final value
+	return map[string]any{storagePath: val}, nil
 }
 
-func replaceIn(path, key, value string) string {
-	// TODO: the PR adding list support to Get() will adapt this to indexes
-	parts := strings.Split(path, ".")
+func replaceIn(path, key, value string) (string, error) {
```

Comment:
the need for this makes one wonder if the related helper shouldn't also carry/traverse a pre-split version of storagePath ?